### PR TITLE
(WIP) Add client to debugger

### PR DESCRIPTION
**Pull request by jasonLaster**
Associated Issue: https://github.com/devtools-html/devtools-core/issues/287

### Summary of Changes

This moves the part of client adapters responsible for debugger commands and events back to debugger. The rationale for this move is:

* other tools like the inspector, console, don't need the debugger client.
* patches to the debugger occasionally touch the client
* versioning and packages is overhead that slows us down
* moving the debugger logic out frees the client adapter to fold into the launchpad and make the launchpad a devtool tab launcher! (find tabs and connect to them)

Yes, this patch scares the hell out of me :)

### CODE REVIEW

File: src/client/chrome/commands.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+function pauseOnExceptions(
+  shouldPauseOnExceptions: boolean,
+  shouldIgnoreCaughtExceptions: boolean
+) {
+  if (!shouldPauseOnExceptions) {
+    return debuggerAgent.setPauseOnExceptions({ state: "none" });
+  }
+  const state = shouldIgnoreCaughtExceptions ? "uncaught" : "all";
```
Comment (nchevobbe) [inline src/client/chrome/commands.js]: Maybe not in the scope of this review, but it would be nice if the `state` possible values (`none`, `uncaught`, `all`, ..) were in constants somewhere

File: src/client/chrome/commands.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  const { result } = await runtimeAgent.getProperties({
+    objectId: object.objectId,
+  });
+
+  const loadedObjects = result.map(createLoadedObject);
+
+  return { loadedObjects };
+}
+
+function evaluate(script: string) {
```
Comment (nchevobbe) [inline src/client/chrome/commands.js]: could we rename `script` to `expression` so it matches what we passe in `evaluate` ?

File: src/client/chrome/commands.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  return runtimeAgent.evaluate({ expression: script });
+}
+
+function debuggeeCommand(script: string) {
+  evaluate(script);
+  return Promise.resolve();
```
Comment (nchevobbe) [inline src/client/chrome/commands.js]: out of curiosity, why are we returning a self-resolving promise here ?

File: src/client/chrome/create.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      line: serverLocation.lineNumber + 1,
+      column: serverLocation.columnNumber,
+    };
+  }
+}
+
+function toServerLocation(location: Location): ServerLocation {
+  return {
+    scriptId: location.sourceId,
+    lineNumber: location.line - 1,
```
Comment (nchevobbe) [inline src/client/chrome/create.js]: shouldn't we have `columnNumber: location.column` too ?

File: src/client/chrome/events.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+  actions.newSource({
+    id: scriptId,
+    url,
+    sourceMapURL,
+    isPrettyPrinted: false,
+  });
+}
+
+function scriptFailedToParse() {}
```
Comment (nchevobbe) [inline src/client/chrome/events.js]: could we add a comment explaining why we must have this ?

File: src/client/chrome/events.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+async function paused(
+  {
+    callFrames,
+    reason,
+    data,
+    hitBreakpoints,
+    asyncStackTrace,
+  }: any
+) {
+  const frames = callFrames.map(createFrame);
```
Comment (nchevobbe) [inline src/client/chrome/events.js]: could we check that callFrames is a non-empty array

File: src/client/chrome/events.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+  const objectId = frame.scopeChain[0].object.objectId;
+  const { result } = await runtimeAgent.getProperties({
+    objectId,
+  });
+
+  const loadedObjects = result.map(createLoadedObject);
+
+  if (clientType == "chrome") {
+    pageAgent.configureOverlay({ message: "Paused in debugger.html" });
```
Comment (nchevobbe) [inline src/client/chrome/events.js]: I guess we should localize the message

File: src/client/chrome/events.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+function resumed() {
+  if (clientType == "chrome") {
+    pageAgent.configureOverlay({ suspended: false });
+  }
+
+  actions.resumed();
+}
+
+function globalObjectCleared() {}
```
Comment (nchevobbe) [inline src/client/chrome/events.js]: add a comment explaining why we need this

File: src/client/chrome/events.js [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+// Page Events
+function frameNavigated(frame: any) {
+  actions.navigated();
+}
+
+function frameStartedLoading() {
+  actions.willNavigate();
+}
+
+function domContentEventFired() {}
```
Comment (nchevobbe) [inline src/client/chrome/events.js]: add a comment explaining why we need those

File: src/client/firefox/events.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+let threadClient: ThreadClient;
+let actions: Actions;
+
+function setupEvents(dependencies: Dependencies) {
+  threadClient = dependencies.threadClient;
+  actions = dependencies.actions;
+
+  Object.keys(clientEvents).forEach(eventName => {
+    if (threadClient) {
```
Comment (nchevobbe) [inline src/client/firefox/events.js]: could we move the forEach call inside the if ? 

```js
if (threadClient) {
  Object.keys(clientEvents).forEach(eventName => {
    threadClient.addListener(eventName, clientEvents[eventName]);
  });
}
```

File: src/client/firefox/types.js [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+ * BreakpointClient
+ * @memberof firefox
+ * @static
+ */
+export type BreakpointClient = {
+  actor: ActorId,
+  remove: () => void,
+  location: Location,
+  setCondition: (ThreadClient, boolean, boolean) => Promise<BreakpointClient>,
+  // getCondition: () => any,
```
Comment (nchevobbe) [inline src/client/firefox/types.js]: why are those commented ?

File: src/components/Editor/index.js [hunk 12]
```diff
@@ -254,11 +254,11 @@ const Editor = React.createClass({
   },
 
   /*
-   * The default Esc command is overridden in the CodeMirror keymap to allow
-   * the Esc keypress event to be catched by the toolbox and trigger the
-   * split console. Restore it here, but preventDefault if and only if there
-   * is a multiselection.
-  */
+     * The default Esc command is overridden in the CodeMirror keymap to allow
```
Comment (nchevobbe) [inline src/components/Editor/index.js]: is this change needed ? I fell like the previous version was correct

File: src/components/Editor/index.js [hunk 13]
```diff
@@ -549,10 +549,10 @@ const Editor = React.createClass({
   },
 
   /**
-   * Handle getting the source document or creating a new
-   * document with the correct mode and text.
-   *
-   */
+     * Handle getting the source document or creating a new
```
Comment (nchevobbe) [inline src/components/Editor/index.js]: is this change needed ? I fell like the previous version was correct

File: assets/panel/panel.js [hunk 14]
```diff
@@ -16,49 +15,45 @@ function DebuggerPanel(iframeWindow, toolbox) {
 }
 
 DebuggerPanel.prototype = {
-  open: Task.async(function* () {
+  open: Task.async(function*() {
```
Comment (jryans) [inline assets/panel/panel.js]: I believe it should be safe to do so here... Firefox supports async / await natively now.

File: assets/panel/panel.js [hunk 15]
```diff
@@ -68,10 +63,10 @@ DebuggerPanel.prototype = {
     return { frames, selected };
   },
 
-  destroy: function () {
+  destroy: function() {
```
Comment (jryans) [inline assets/panel/panel.js]: DevTools ESLint style in m-c requires `function (` with the space. Should we update the debugger `.eslintrc` so that they match on this point?

Comment (jasonLaster) [inline assets/panel/panel.js]: oops, this change was accidental. 

We could update the eslint to be consistent, but we started using prettier on the debugger for the styling and I like the defaults that it has. (even though it's not consistent with devtools in this case)

**Comment by jasonLaster** [comment]
@nchevobbe many of your comments were on chrome/firefox files that were copied over. I'll reply, but we shouldn't change these things now

File: src/client/firefox/types.js (hunk 11 above)
Comment (jasonLaster) [inline src/client/firefox/types.js]: hmm, i don't remember.

File: src/client/chrome/events.js (hunk 7 above)
Comment (jasonLaster) [inline src/client/chrome/events.js]: it's a chrome message, but yea

File: src/client/chrome/create.js (hunk 4 above)
Comment (jasonLaster) [inline src/client/chrome/create.js]: yep. good point. this will break w/ source maps

File: src/client/chrome/commands.js (hunk 3 above)
Comment (jasonLaster) [inline src/client/chrome/commands.js]: yep :) 

It's a weird function that's only for testing.

File: src/client/chrome/commands.js (hunk 2 above)
Comment (jasonLaster) [inline src/client/chrome/commands.js]: yeah - that would be clearer

**Comment by nchevobbe** [comment]
@jasonLaster sure
